### PR TITLE
[refactor-csv-processing] switch to csv per-line processing

### PR DESCRIPTION
**Pull request by yyakovlev**
[refactor-csv-processing] switch to csv per-line processing

### CODE REVIEW

File: scan-project/sample/config/scan-application.yml [hunk 1]
```diff
@@ -1,7 +1,7 @@
 read:
   # !!! Project folder to be scanned - provide actual value !!!
-  folder: "/work/project/scanned-sample"
-
+  #folder: "/work/project/scanned-sample"
+  folder: "C:/Work/projects/code-egress-sample-scan"
```
Comment (kbabkin) [inline scan-project/sample/config/scan-application.yml]: Local change?

File: src/main/java/com/bt/code/egress/process/WordReplacer.java [hunk 2]
```diff
@@ -52,6 +52,10 @@ public String replace(WordMatch wordMatch) {
         }
         String predefined = wordMatch.getTemplate();
         String template = StringUtils.isBlank(predefined) ? defaultTemplate : predefined;
+
+        if (wordMatch.getReplacement() != null) {
```
Comment (kbabkin) [inline src/main/java/com/bt/code/egress/process/WordReplacer.java]: Can be done as first step

File: src/main/java/com/bt/code/egress/process/LineReplacer.java [hunk 3]
```diff
@@ -53,7 +62,7 @@ public String replace(String line, LineLocation lineLocation, TextMatched.Listen
             WordMatch wordMatch = matchParam.getWordMatch();
             LineToken lineToken = wordMatch.getLineToken();
 
-            String replacement = wordReplacer.replace(wordMatch);
+            String replacement = wordMatch.getReplacement() != null ? wordMatch.getReplacement() : wordReplacer.replace(wordMatch);
```
Comment (kbabkin) [inline src/main/java/com/bt/code/egress/process/LineReplacer.java]: Left it to wordReplacer

File: src/main/java/com/bt/code/egress/process/FileReplacer.java [hunk 4]
```diff
@@ -20,18 +19,35 @@
...
+        boolean isCsv = isEligibleForCsvReplacement(file.getFilename());
+        if (isCsv) {
+            log.info("CSV processing will be applied for {}", file);
+        }
+
         String line;
         int lineNum = 0;
         ArrayList<String> originalLines = new ArrayList<>();
         List<String> replacedLines = new ArrayList<>();
+        int lineNumber = 0;
```
Comment (kbabkin) [inline src/main/java/com/bt/code/egress/process/FileReplacer.java]: Duplicates lineNum ?

File: src/main/java/com/bt/code/egress/read/LineToken.java [hunk 5]
```diff
@@ -1,9 +1,11 @@
 package com.bt.code.egress.read;
 
+import lombok.EqualsAndHashCode;
 import lombok.Getter;
 import lombok.RequiredArgsConstructor;
 
 @RequiredArgsConstructor
+@EqualsAndHashCode
```
Comment (kbabkin) [inline src/main/java/com/bt/code/egress/read/LineToken.java]: Is it needed? should not include cached values like wordLowerCase

**Review by kbabkin — APPROVED** [review]
Integration with project is OK, have not deeply checked new CSV* files

File: src/main/java/com/bt/code/egress/read/CsvLineMatcher.java [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+                                  String line) {
+        Collection<WordMatch> allMatches = union(union(replacements, cleanups), fills);
+        if (CollectionUtils.isEmpty(allMatches)) {
+            return Collections.emptyList();
+        }
+        LinkedList<WordMatch> allMatchesSorted = new LinkedList<>(allMatches);
+        allMatchesSorted.sort(Comparator.comparing(m -> m.getLineToken().getStartPos()));
+        //ha ha who is sure where linked list's last element is ? -)
+        int startPos = allMatchesSorted.getFirst().getLineToken().getStartPos();
+        int endPos = allMatchesSorted.getLast().getLineToken().getEndPos();
```
Comment (kbabkin) [inline src/main/java/com/bt/code/egress/read/CsvLineMatcher.java]: Is it last? sorted by startPos, not endPos

**Review by d57344 — APPROVED** [review]

File: src/main/java/com/bt/code/egress/read/LineToken.java (hunk 5 above)
Comment (kbabkin) [reply src/main/java/com/bt/code/egress/read/LineToken.java]: Fixed

File: src/main/java/com/bt/code/egress/App.java [hunk 7]
```diff
@@ -82,9 +91,29 @@ public void run(ApplicationArguments args) {
 
         wordReplacer.saveGenerated(writeGeneratedReplacement.toPath());
 
+        logErrors(fileReplacer);
```
Comment (kbabkin) [inline src/main/java/com/bt/code/egress/App.java]: can it be moved to fileReplacer? so in App only something like ``fileReplacer.verify()``

File: src/main/java/com/bt/code/egress/process/FileReplacer.java (hunk 4 above)
Comment (d57344) [reply src/main/java/com/bt/code/egress/process/FileReplacer.java]: .. and below 
 ++lineNum
 lineNumber++

File: scan-project/sample/config/scan-application.yml (hunk 1 above)
Comment (yyakovlev) [reply scan-project/sample/config/scan-application.yml]: I see it has been removed

File: src/main/java/com/bt/code/egress/process/WordReplacer.java (hunk 2 above)
Comment (yyakovlev) [reply src/main/java/com/bt/code/egress/process/WordReplacer.java]: Fixed

File: src/main/java/com/bt/code/egress/process/LineReplacer.java (hunk 3 above)
Comment (yyakovlev) [reply src/main/java/com/bt/code/egress/process/LineReplacer.java]: Fixed

File: src/main/java/com/bt/code/egress/process/FileReplacer.java (hunk 4 above)
Comment (yyakovlev) [reply src/main/java/com/bt/code/egress/process/FileReplacer.java]: Fixed

File: src/main/java/com/bt/code/egress/read/CsvLineMatcher.java (hunk 6 above)
Comment (yyakovlev) [reply src/main/java/com/bt/code/egress/read/CsvLineMatcher.java]: CSV replacements config should provide consistent, non-overlapping matches, that produce the same order by startPos and by endPos. I've added code that removes matches that violate that rule

File: src/main/java/com/bt/code/egress/App.java (hunk 7 above)
Comment (yyakovlev) [reply src/main/java/com/bt/code/egress/App.java]: Moved

File: scan-project/sample/config/scan-application.yml (hunk 1 above)
Comment (kbabkin) [reply scan-project/sample/config/scan-application.yml]: No, it is still here

File: src/main/java/com/bt/code/egress/App.java (hunk 7 above)
Comment (kbabkin) [reply src/main/java/com/bt/code/egress/App.java]: So ``logErrors`` method can be removed

Comment (yyakovlev) [reply src/main/java/com/bt/code/egress/App.java]: Sorry, missed that. Method removed

File: scan-project/sample/config/scan-application.yml (hunk 1 above)
Comment (yyakovlev) [reply scan-project/sample/config/scan-application.yml]: Ah I was looking at the other file. So now finally reverted

**Pull request merged**